### ISSUE TITLE
Running composer script during dyno boot crashes application

### ISSUE THREAD

**Issue by holtkamp** [heroku/heroku-buildpack-php#354]
As of today the following error is displayed when deploying an application:

```
2019-08-07T09:43:53.005605+00:00 app[web.1]: [heroku-exec] Starting
2019-08-07T09:43:53.006777+00:00 app[web.1]: warming up PHP OpCode caches for application: MyApplicationName
2019-08-07T09:43:54.164455+00:00 heroku[web.1]: State changed from starting to crashed
2019-08-07T09:43:54.236785+00:00 heroku[web.1]: State changed from crashed to starting
2019-08-07T09:43:54.113844+00:00 app[web.1]: Unable to determine Composer vendor-dir setting; is 'composer' executable on path or 'composer.phar' in current working directory?
2019-08-07T09:43:54.151332+00:00 heroku[web.1]: Process exited with status 1

```

In `Procfile` we have:

```bash
web: ./startWebProcess.sh
```

Then in `startWebProcess.sh`:

```bash
# Manually warm-up the OpCode cache to prevent cache slams / thundering herd problem
# @link https://tideways.com/profiler/blog/dodge-the-thundering-herd-with-file-based-opcache-in-php7
# Warming up the OpCode caches takes around 30 / 40 seconds. When the WebServer has not started within 25 seconds
# Heroku assumes it is crashed. Therefore we should use the ampersand "&" to start this process in the background.
# This way the WebServer does not have to wait for this process to end
echo "warming up PHP OpCode caches for application:" $HEROKU_APP_NAME
composer warmup-opcode &

# Eventually: launch the Apache WebServer which has PHP support and "public" as WebRoot
vendor/bin/heroku-php-apache2 public/
```

This way of invoking the `composer warmup-opcode` script (see https://github.com/jderusse/composer-warmup) used to work for quite some time.

Versions:

```
 -----> Installing platform packages...        
remote:        - php (7.3.8)        
remote:        - ext-gd (bundled with php)        
remote:        - ext-intl (bundled with php)        
remote:        - ext-mbstring (bundled with php)        
remote:        - ext-soap (bundled with php)        
remote:        - ext-imagick (3.4.4)        
remote:        - ext-apcu (5.1.17)        
remote:        - apache (2.4.39)        
remote:        - nginx (1.16.0)        
remote: -----> Installing dependencies...        
remote:        Composer version 1.9.0 2019-08-02 20:55:32  
```

Might be related to: https://github.com/heroku/heroku-buildpack-php/issues/348 and https://github.com/heroku/heroku-buildpack-php/pull/351 ?

Is it still possible to invoke a `composer` script? And if so, how to do it correctly?

Note that using `heroku run bash --app my-application-name` and then invoke  `composer warmup-opcode` works properly...

**UPDATE**
Apparently the error message itself comes from running `vendor/bin/heroku-php-apache2 public/`. This might have been there for a while, but apparently combined with another command it now crashes the application. 

Manually reverting the used buildpack to [`v158`](https://github.com/heroku/heroku-buildpack-php/releases/tag/v158) resolves the problem with the crashing app:

```
heroku buildpacks:set https://github.com/heroku/heroku-buildpack-php\#v158 -a my-application-name
```

So maybe the new version `v159` is a bit "stricter" in dealing with error messages and marking the app as being crashed?

**Comment by holtkamp** [comment]
**UPDATE**
Apparently the error itself comes from running `vendor/bin/heroku-php-apache2 public/`. This might have been there for a while, but apparently combined with another command it now crashes the application.

**Comment by dzuelke** [comment]
Thanks for the report. Will investigate.

To clarify, since the ticket title says "during deploy": this happens on dyno boot, not during deploy, correct?

**Comment by holtkamp** [comment]
Indeed: when trying to boot the "web" process of the Profile.

"Deploy" is probably something else in the Heroku domain ;)

**Comment by dzuelke** [comment]
Hmh... this doesn't work as a reproduce case yet. How are you setting `opcache.file_cache`?

Stuff like this is generally much easier to debug if you open a support ticket, because then we can look at your app directly.

**Comment by dzuelke** [comment]
Even if I fiddle something together, I can't reproduce this yet. The app boots fine.

That plugin however is broken:

```
2019-08-08T01:17:46.677383+00:00 app[web.1]: Fatal error: Cannot declare interface Jderusse\Warmup\ClassmapReader\ReaderInterface, because the name is already in use in /app/vendor/jderusse/composer-warmup/src/ClassmapReader/ReaderInterface.php on line 7
```

But that doesn't matter for the purpose of reproducing this.

Please open a ticket and say I asked for immediate escalation.

**Comment by dzuelke** [comment]
(and yeah, a "deploy" is when you `git push heroku master` on Planet Heroku ;) )

**Comment by dzuelke** [comment]
You're probably doing some stuff either early in your `boot.sh`, or in a `.profile` script, right? To create the dir, somehow inject the INI stuff, etc? I'd need to see all of that please if you want to share it here.

**Comment by holtkamp** [comment]
Thanks for diving into this:

>To clarify, since the ticket title says "during deploy": this happens on dyno boot, not during deploy, correct?

I changed the issue title to underline that it is about "dyno boot" phase

>That plugin however is broken:
`2019-08-08T01:17:46.677383+00:00 app[web.1]: Fatal error: Cannot declare interface Jderusse\Warmup\ClassmapReader\ReaderInterface, because the name is already in use in /app/vendor/jderusse/composer-warmup/src/ClassmapReader/ReaderInterface.php on line 7`

Ah, yeah, I realized I forgot to mention that: it "is" possible that the composer plugin generates errors, for example when a class is already in use, or a reserved keyword is used as classname, for example "Object". (Maybe Heroku now "detects" these errors and assumes booting the application has failed?)

>How are you setting opcache.file_cache?

Inspired by the [plugin documentation](https://github.com/jderusse/composer-warmup#configure) I use: 
```ini
#/app/.heroku/php/etc/php/conf.d/opcache.ini
[OPcache]
opcache.enable = 1
opcache.enable_cli = 1
opcache.enable_file_override = 1
opcache.fast_shutdown = 1
opcache.file_cache = /tmp
opcache.file_cache_consistency_checks = 1
opcache.file_update_protection = 0
opcache.huge_code_pages = 1
opcache.interned_strings_buffer = 16
opcache.max_accelerated_files = 32531
opcache.memory_consumption = 196
opcache.save_comments = 1
opcache.validate_timestamps = 0
```

>You're probably doing some stuff either early in your boot.sh, or in a .profile script, right? To create the dir, somehow inject the INI stuff, etc? I'd need to see all of that please if you want to share it here.

Mmm, AFAIK I am not doing any exotic stuff like creating a dir / inject INI stuff. The idea is "just" to warmup the opcode caches before starting the webserver. As described in this [Twitter thread](https://twitter.com/matthieunapoli/status/987798656349868033), I tried "warming up opcode" cache during the "compile" step, but this will not work. 

Offtopic: maybe this use case indicates the "wish" for a "boot" step in `composer.json`. 

>Stuff like this is generally much easier to debug if you open a support ticket, because then we can look at your app directly.

Ok, will do this today.

Thanks again!

**Comment by dzuelke** [comment]
@holtkamp the sooner the better, please. No need for anything complicated.

Make a ticket that says "PHP boot crash regression", write a body like "Hi support, dzuelke asked me to open this ticket ASAP so you can escalate it to him for investigation, see GitHub: https://github.com/heroku/heroku-buildpack-php/issues/354", done.

What matter is that you select the right issue type, pick the affected app from the list, and check the box that allows us access for investigation.

**Comment by holtkamp** [comment]
@dzuelke just created a Heroku support ticket:

`#754318 - PHP boot crash regression: running composer script during dyno boot crashes application on Heroku buildpack v159`

**Comment by dzuelke** [comment]
Fun times. For anyone interested, this is due to OPcache file caching enabled, which obviously also caches opcodes generated by the New Relic extension, but then if PHP is run with `-dnewrelic.enabled=0`, stuff blows up:

```
$ php -v
PHP 7.3.8 (cli) (built: Aug  6 2019 01:58:23) ( NTS )
$ php -m | grep -i -e newrelic -e opcache
newrelic
Zend OPcache
Zend OPcache

$ composer require jderusse/composer-warmup
Using version ^1.3 for jderusse/composer-warmup
…

$ mkdir /tmp/opcache

$ php -dopcache.enable_cli=1 -dopcache.file_cache=/tmp/opcache $(which
composer) warmup-opcode
2019-08-10 02:28:49.598 +0000 (41 41) info: New Relic 8.7.0.242 …
…
(a bunch of fatal errors, those are expected, just ignore)

$ php -dopcache.enable_cli=1 -dopcache.file_cache=/tmp/opcache $(which
composer) config vendor-dir
2019-08-10 02:28:49.598 +0000 (41 41) info: New Relic 8.7.0.242 …
vendor

$ php -dnewrelic.enabled=0 -dopcache.enable_cli=1
-dopcache.file_cache=/tmp/opcache $(which composer) config vendor-dir
2019-08-10 02:29:01.004 +0000 (54 54) info: New Relic PHP Agent
globally disabled
Segmentation fault
```

BUT!

If you run the warmup-opcode step with `-dnewrelic.enabled=0` as well,
then nothing fails:

```
$ rm -rf /tmp/opcache/*

$ php -dnewrelic.enabled=0 -dopcache.enable_cli=1
-dopcache.file_cache=/tmp/opcache $(which composer) warmup-opcode
2019-08-10 02:33:57.016 +0000 (8 8) info: New Relic PHP Agent globally disabled
You should set the `opcache.file_update_protection` to 0 in order to
cache recently updated files
…

$ php -dopcache.enable_cli=1 -dopcache.file_cache=/tmp/opcache $(which
composer) config vendor-dir
2019-08-10 02:34:09.879 +0000 (43 43) info: New Relic 8.7.0.242 …
vendor

$ php -dnewrelic.enabled=0 -dopcache.enable_cli=1
-dopcache.file_cache=/tmp/opcache $(which composer) config vendor-dir
2019-08-10 02:34:14.937 +0000 (56 56) info: New Relic PHP Agent
globally disabled
vendor
```

**Comment by dzuelke** [comment]
(IIRC this was fixed in a 9.0.x `ext-newrelic` point update)
